pkg/kafka/partition/reader_refactor.go

```diff
@@ -0,0 +1,289 @@
+		}),
+		fetchesTotal: promauto.With(r).NewCounter(prometheus.CounterOpts{
+			Name: "loki_kafka_reader_fetches_total",
+			Help: "Total number of Kafka fetches performed.",
+		}),
+		receiveDelay: promauto.With(r).NewHistogram(prometheus.HistogramOpts{
+			Name:                            "loki_kafka_reader_receive_delay_seconds",
+			Help:                            "Delay between producing a record and receiving it.",
+			NativeHistogramZeroThreshold:    math.Pow(2, -10),
+			NativeHistogramBucketFactor:     1.2,
+			NativeHistogramMaxBucketNumber:  100,
+			NativeHistogramMinResetDuration: 1 * time.Hour,
+			Buckets:                         prometheus.ExponentialBuckets(0.125, 2, 18),
+		}),
+	}
+}
+
+// Reader provides low-level access to Kafka partition reading operations
+type RefactoredReader struct {
+	client        *kgo.Client
+	topic         string
+	partitionID   int32
+	consumerGroup string
+	metrics       *refactoredReaderMetrics
+	logger        log.Logger
+}
+
+// NewReader creates a new Reader instance
+func NewRefactoredReader(
+	client *kgo.Client,
+	topic string,
+	partitionID int32,
+	consumerGroup string,
+	logger log.Logger,
+	reg prometheus.Registerer,
+) *RefactoredReader {
+	return &RefactoredReader{
+		client:        client,
+		topic:         topic,
+		partitionID:   partitionID,
+		consumerGroup: consumerGroup,
+		metrics:       newRefactoredReaderMetrics(reg),
+		logger:        logger,
+	}
+}
+
+// Topic returns the topic being read
+func (r *RefactoredReader) Topic() string {
+	return r.topic
+}
+
+// Partition returns the partition being read
+func (r *RefactoredReader) Partition() int32 {
+	return r.partitionID
+}
+
+// ConsumerGroup returns the consumer group
+func (r *RefactoredReader) ConsumerGroup() string {
+	return r.consumerGroup
+}
+
+// FetchLastCommittedOffset retrieves the last committed offset for this partition
+func (r *RefactoredReader) FetchLastCommittedOffset(ctx context.Context) (int64, error) {
+	req := kmsg.NewPtrOffsetFetchRequest()
+	req.Topics = []kmsg.OffsetFetchRequestTopic{{
+		Topic:      r.topic,
+		Partitions: []int32{r.partitionID},
+	}}
+	req.Group = r.consumerGroup
+
+	resps := r.client.RequestSharded(ctx, req)
+
+	// Since we issued a request for only 1 partition, we expect exactly 1 response.
+	if expected, actual := 1, len(resps); actual != expected {
+		return 0, fmt.Errorf("unexpected number of responses: %d", len(resps))
+	}
+
+	// Ensure no error occurred.
+	res := resps[0]
+	if res.Err != nil {
+		return 0, res.Err
+	}
+
+	// Parse the response.
+	fetchRes, ok := res.Resp.(*kmsg.OffsetFetchResponse)
+	if !ok {
+		return 0, errors.New("unexpected response type")
+	}
+
+	if len(fetchRes.Groups) != 1 ||
+		len(fetchRes.Groups[0].Topics) != 1 ||
+		len(fetchRes.Groups[0].Topics[0].Partitions) != 1 {
+		level.Debug(r.logger).Log(
+			"msg", "malformed response, setting to start offset",
+		)
+		return int64(KafkaStartOffset), nil
```

Comment:
like the other failures cases in this method, can we also return error here and let the caller decide on the offset?

Reply:
I agree with you, but my goal here is to refactor the existing implementation, not change it's logic. This is a good followup PR though.